File: iCyption2022/CftTest/Escape/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using vb = vector<bool>;
using vvb = vector<vb>;

void solve() {
  int n;
  cin >> n;
  int arr[n][n];
  for (int i = 0; i < n; i++)
    for (int j = 0; j <= i; j++)
      cin >> arr[i][j];

  vvb reachable(n, vb(n, false));

  reachable[0][0] = true;
  for (int i = 0; i < n - 1; i++) {
    for (int j = 0; j <= i; j++) {
      if (reachable[i][j]) {
        if (arr[i][j] >= arr[i+1][j])
          reachable[i+1][j] = true;
        if (arr[i][j] >= arr[i+1][j+1])
          reachable[i+1][j+1] = true;
      }
    }
  }

  bool bisa = false;
  for (int j = 0; j < n; j++)
    if (reachable[n-1][j])
      bisa = true;

  cout << (bisa ? "BISA" : "MUSTAHIL") << '\n';
}
 
int main() {
  ios::sync_with_stdio(0), cin.tie(0);
  solve();
}

File: iCyption2022/CftTest/Escape/solution.py
n = int(input())

arr = [[0 for i in range(n)] for j in range(n)]

for i in range(n) :
    arr[i] = list(map(int, input().split()))
    for j in range(n-i): 
        arr[i].append(0)

reachable = [[False for i in range(n)] for j in range(n)]

reachable[0][0] = True;
for i in range(n-1):
    for j in range(n):
        if (reachable[i][j]):
            if (arr[i][j] >= arr[i+1][j]):
                reachable[i+1][j] = True
            if (arr[i][j] >= arr[i+1][j+1]):
                reachable[i+1][j+1] = True

bisa = False
for j in range(n):
    if (reachable[n-1][j]):
        bisa = True

if (bisa):
    print("BISA") 
else:
    print("MUSTAHIL")
